api/src/test/java/org/pmiops/workbench/billing/BillingAlertsServiceTest.java

```diff
@@ -0,0 +1,195 @@
+    workbenchConfig = WorkbenchConfig.createEmptyConfig();
+
+    InputStream inputStream =
+        getClass().getClassLoader().getResourceAsStream("bigquery/get_billing_project_costs.ser");
+    TableResult tableResult = (TableResult) (new ObjectInputStream(inputStream)).readObject();
+
+    doReturn(tableResult).when(bigQueryService).executeQuery(any());
+  }
+
+  @Test
+  public void findCreatorByWorkspaceNamespace() {
+    User user = createUser("test@test.com");
+    createWorkspace(user, "rumney");
+    createWorkspace(user, "rumney");
+
+    assertThat(userDao.findCreatorByWorkspaceNamespace("rumney")).isEqualTo(user);
+  }
+
+  @Test
+  public void alertUsersExceedingFreeTierBilling_singleProjectExceedsLimit() {
+    workbenchConfig.billing.defaultFreeCreditsLimit = 100.0;
+
+    User user = createUser("test@test.com");
+    createWorkspace(user, "aou-test-f1-26");
+
+    billingAlertsService.alertUsersExceedingFreeTierBilling();
+    verify(notificationService).alertUser(eq(user), any());
+  }
+
+  @Test
+  public void alertUsersExceedingFreeTierBilling_deletedUserNotIgnored() {
+    workbenchConfig.billing.defaultFreeCreditsLimit = 100.0;
+
+    User user = createUser("test@test.com");
+    user.setDisabled(true);
+    userDao.save(user);
+    createWorkspace(user, "aou-test-f1-26");
+
+    billingAlertsService.alertUsersExceedingFreeTierBilling();
+    verify(notificationService).alertUser(eq(user), any());
+  }
+
+  @Test
+  public void alertUsersExceedingFreeTierBilling_deletedWorkspaceNotIgnored() {
+    workbenchConfig.billing.defaultFreeCreditsLimit = 100.0;
+
+    User user = createUser("test@test.com");
+    Workspace workspace = createWorkspace(user, "aou-test-f1-26");
+    workspace.setWorkspaceActiveStatusEnum(WorkspaceActiveStatus.DELETED);
+    workspaceDao.save(workspace);
+
+    billingAlertsService.alertUsersExceedingFreeTierBilling();
+    verify(notificationService).alertUser(eq(user), any());
+  }
+
+  @Test
+  public void alertUsersExceedingFreeTierBilling_noAlert() {
+    workbenchConfig.billing.defaultFreeCreditsLimit = 500.0;
+
+    User user = createUser("test@test.com");
+    createWorkspace(user, "aou-test-f1-26");
+
+    billingAlertsService.alertUsersExceedingFreeTierBilling();
+    verifyZeroInteractions(notificationService);
+  }
+
+  @Test
+  public void alertUsersExceedingFreeTierBilling_workspaceMissingCreator() {
+    workbenchConfig.billing.defaultFreeCreditsLimit = 500.0;
+
+    User user = createUser("test@test.com");
+    createWorkspace(user, "aou-test-f1-26");
+    createWorkspace(null, "rumney");
+
+    billingAlertsService.alertUsersExceedingFreeTierBilling();
+    verifyZeroInteractions(notificationService);
+  }
+
+  @Test
+  public void alertUsersExceedingFreeTierBilling_combinedProjectsExceedsLimit() {
+    workbenchConfig.billing.defaultFreeCreditsLimit = 500.0;
+
+    User user = createUser("test@test.com");
+    createWorkspace(user, "aou-test-f1-26");
```

Comment:
Hmm, the downside to this serialization approach is that I have no way of directly inspecting the test data in source code. e.g. what is the value for each of these workspaces? Presumably it sums to >500, but if this test broke I'd be a bit lost. Even something like JSON checked in I think would be a step up from this binary serialization.

Reply:
I looked into some methods to generate the TableResult from the JSON but couldn't find a quick way where I would trust the generated object's consistency with the real thing. I checked in the JSON as an intermediary measure.